File: benchmarks/fixtures/webapp_rb/api/v2/payments.rb
# API v2 payment endpoints with webhook support.

require_relative '../../utils/helpers'
require_relative '../../validators/payment'
require_relative '../../services/payment/processor'
require_relative '../../database/connection'
require_relative '../../events/dispatcher'

module ApiV2Payments
  # Create payment with idempotency.
  def self.handle_create_payment(request, db, events)
    validate_request(request)
    body = request[:body]
    headers = request[:headers] || {}
    idempotency_key = headers['Idempotency-Key'] || ''
    get_logger('api.v2.payments').info("V2 create payment (idempotency=#{idempotency_key[0, 12]})")
    payment_data = PaymentValidator.validate(body)
    processor = PaymentProcessor.new(db, events)
    result = processor.process_payment(
      payment_data[:user_id],
      payment_data[:amount],
      payment_data[:currency]
    )
    { status: 201, data: result }
  end

  # Handle webhook.
  def self.handle_webhook(request, db, events)
    validate_request(request)
    body = request[:body]
    event_type = body[:type]
    get_logger('api.v2.payments').info("Webhook: #{event_type}")
    { status: 200, data: { acknowledged: true } }
  end

  # Revenue report.
  def self.handle_revenue_report(request, db, events)
    validate_request(request)
    processor = PaymentProcessor.new(db, events)
    { status: 200, data: {} }
  end
end

File: benchmarks/fixtures/webapp_ts/src/api/v2/payments.ts
/** API v2 payment endpoints with webhook support. */
import { getLogger, validateRequest } from '../../utils/helpers';
import { validate as validatePayment } from '../../validators/payment';
import { PaymentProcessor } from '../../services/payment/processor';
import { DatabaseConnection } from '../../database/connection';
import { EventDispatcher } from '../../events/dispatcher';

const logger = getLogger("api.v2.payments");

/** Create payment with idempotency. */
export async function handleCreatePayment(request: Record<string, unknown>, db: DatabaseConnection, events: EventDispatcher): Promise<Record<string, unknown>> {
    validateRequest(request);
    const body = request["body"] as Record<string, unknown>;
    const headers = request["headers"] as Record<string, string>;
    const idempotencyKey = headers?.["Idempotency-Key"] ?? "";
    logger.info(`V2 create payment (idempotency=${idempotencyKey.substring(0, 12)})`);
    const paymentData = validatePayment(body);
    const processor = new PaymentProcessor(db, events);
    processor.initialize();
    const result = await processor.processPayment(
        paymentData["user_id"] as string,
        paymentData["amount"] as number,
        paymentData["currency"] as string,
    );
    return { status: 201, data: result };
}

/** Handle webhook. */
export async function handleWebhook(request: Record<string, unknown>, db: DatabaseConnection, events: EventDispatcher): Promise<Record<string, unknown>> {
    validateRequest(request);
    const body = request["body"] as Record<string, unknown>;
    const eventType = body["type"] as string;
    logger.info(`Webhook: ${eventType}`);
    return { status: 200, data: { acknowledged: true } };
}

/** Revenue report. */
export async function handleRevenueReport(request: Record<string, unknown>, db: DatabaseConnection, events: EventDispatcher): Promise<Record<string, unknown>> {
    validateRequest(request);
    const processor = new PaymentProcessor(db, events);
    processor.initialize();
    return { status: 200, data: {} };
}
